Drop String::characters8() / characters16()

cdumez opened the pull request:
#### ca2f1e056444073522e176a7bfb47aeeabacbee9
<pre>
Drop String::characters8() / characters16()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273352">https://bugs.webkit.org/show_bug.cgi?id=273352</a>

Reviewed by Sam Weinig and Darin Adler.

Drop String::characters8() / characters16(), in favor of span equivalents.

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::~OpaqueJSString):
* Source/JavaScriptCore/API/OpaqueJSString.h:
(OpaqueJSString::OpaqueJSString):
* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::ListFormatInput::ListFormatInput):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier&lt;CharType&gt;::append):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings):
* Source/JavaScriptCore/runtime/RegExpConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/RegExpInlines.h:
(JSC::RegExp::matchInline):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
(JSC::codePointAt):
* Source/JavaScriptCore/runtime/TemplateObjectDescriptor.h:
(JSC::TemplateObjectDescriptor::calculateHash):
* Source/JavaScriptCore/yarr/YarrJIT.h:
(JSC::Yarr::YarrCodeBlock::execute):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::copyBaseWindowsDriveLetter):
(WTF::URLParser::copyASCIIStringUntil):
(WTF::URLParser::URLParser):
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::reserveCapacity):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::characters const):
* Source/WTF/wtf/text/StringBuilderJSON.cpp:
(WTF::appendQuotedJSONStringInternal):
(WTF::StringBuilder::appendQuotedJSONString):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::characters&lt;LChar&gt; const): Deleted.
(WTF::String::characters&lt;UChar&gt; const): Deleted.
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
* Source/WebCore/dom/make_names.pl:
(printTagNameCppFile):
(printNodeNameCppFile):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::appendCharactersReplacingEntitiesInternal):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::containsKanaLetters):
* Source/WebCore/html/track/VTTScanner.cpp:
(WebCore::VTTScanner::VTTScanner):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::fixUnparsedProperties):
(WebCore::StyleSheetHandler::fixUnparsedPropertyRanges):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::replaceNonPreservedNewLineAndTabCharactersAndAppend):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::firstUserPerceivedCharacterLength):
* Source/WebCore/rendering/LegacyInlineIterator.h:
(WebCore::LegacyInlineIterator::incrementByCodePointInTextNode):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::xmlDocPtrForString):
* Source/WebGPU/WGSL/Lexer.h:
(WGSL::Lexer::Lexer):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.cpp:
(WebKit::NetworkCache::hexDigitsToHash):
(WebKit::NetworkCache::Key::stringToHash):
* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;String&gt;::encode):
* Source/WebKit/Shared/API/c/cf/WKStringCF.mm:
(WKStringCopyCFString):

Canonical link: <a href="https://commits.webkit.org/278082@main">https://commits.webkit.org/278082@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27850aea1e117a595abf6f7d0fccce63361b1611

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49428 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28710 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52667 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/101 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51732 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34726 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26330 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40351 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51528 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26258 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21468 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43769 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7797 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/42738 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45631 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44274 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54179 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/48921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20700 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47713 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42790 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46722 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26621 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56411 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7106 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25505 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11586 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -321,15 +321,15 @@ static ALWAYS_INLINE JSString* jsSpliceSubstrings(JSGlobalObject* globalObject,
         Checked<int, AssertNoOverflow> bufferPos = 0;
         for (int i = 0; i < rangeCount; i++) {
             int srcLen = substringRanges[i].distance();
```
Change this to `size_t` to get rid of the cast below?

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -321,15 +321,15 @@ static ALWAYS_INLINE JSString* jsSpliceSubstrings(JSGlobalObject* globalObject,
         Checked<int, AssertNoOverflow> bufferPos = 0;
```
This should be `size_t`.

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -321,15 +321,15 @@ static ALWAYS_INLINE JSString* jsSpliceSubstrings(JSGlobalObject* globalObject,
         Checked<int, AssertNoOverflow> bufferPos = 0;
         for (int i = 0; i < rangeCount; i++) {
```
This should be a range-based for loop.

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -340,7 +340,7 @@ static ALWAYS_INLINE JSString* jsSpliceSubstrings(JSGlobalObject* globalObject,
     Checked<int, AssertNoOverflow> bufferPos = 0;
```
`size_t`

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -340,7 +340,7 @@ static ALWAYS_INLINE JSString* jsSpliceSubstrings(JSGlobalObject* globalObject,
     Checked<int, AssertNoOverflow> bufferPos = 0;
     for (int i = 0; i < rangeCount; i++) {
```
range-based

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -340,7 +340,7 @@ static ALWAYS_INLINE JSString* jsSpliceSubstrings(JSGlobalObject* globalObject,
     Checked<int, AssertNoOverflow> bufferPos = 0;
     for (int i = 0; i < rangeCount; i++) {
         int srcLen = substringRanges[i].distance();
-        StringImpl::copyCharacters(buffer + bufferPos.value(), { sourceData + substringRanges[i].begin(), static_cast<size_t>(srcLen) });
```
`size_t`

darinadler commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -1028,9 +1028,10 @@ static inline char32_t codePointAt(const String& string, unsigned position, unsi
+        return string.span8()[position];
     char32_t character;
-    U16_NEXT(string.characters16(), position, length, character);
+    auto characters = string.span16();
```
Why the local variable?

cdumez replied:
Didn't we establish that the local variable was needed or it fails building on Mac-ews? (though it builds fine locally).
__u2 unused variable. You said last time that it was fixed but it still doesn't work on EWS for some reason.

darinadler commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -440,15 +440,15 @@ bool URLParser::copyBaseWindowsDriveLetter(const URL& base)
         if (base.m_string.is8Bit()) {
-            const LChar* begin = base.m_string.characters8();
-            CodePointIterator<LChar> c({ begin + base.m_hostEnd + base.m_portLength + 1, begin + base.m_string.length() });
+            auto characters = base.m_string.span8();
```
Why the local variable?

darinadler commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -440,15 +440,15 @@ bool URLParser::copyBaseWindowsDriveLetter(const URL& base)
         } else {
-            const UChar* begin = base.m_string.characters16();
-            CodePointIterator<UChar> c({ begin + base.m_hostEnd + base.m_portLength + 1, begin + base.m_string.length() });
+            auto characters = base.m_string.span16();
```
Why the local variable?

darinadler commented on Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
```diff
@@ -44,15 +44,13 @@ void ArgumentCoder<String>::encode(Encoder& encoder, const String& string)
     bool is8Bit = string.is8Bit();
-
-    encoder << length << is8Bit;
+    encoder << string.length() << is8Bit;
```
I'm surprised that `encodeSpan` does not encode the length.

cdumez replied:
It currently doesn't for some reason. However, it does seem like it probably should.

cdumez committed:
Drop String::characters8() / characters16()
https://bugs.webkit.org/show_bug.cgi?id=273352

Reviewed by Sam Weinig and Darin Adler.

Drop String::characters8() / characters16(), in favor of span equivalents.

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::~OpaqueJSString):
* Source/JavaScriptCore/API/OpaqueJSString.h:
(OpaqueJSString::OpaqueJSString):
* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::ListFormatInput::ListFormatInput):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier<CharType>::append):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings):
* Source/JavaScriptCore/runtime/RegExpConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/RegExpInlines.h:
(JSC::RegExp::matchInline):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
(JSC::codePointAt):
* Source/JavaScriptCore/runtime/TemplateObjectDescriptor.h:
(JSC::TemplateObjectDescriptor::calculateHash):
* Source/JavaScriptCore/yarr/YarrJIT.h:
(JSC::Yarr::YarrCodeBlock::execute):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::copyBaseWindowsDriveLetter):
(WTF::URLParser::copyASCIIStringUntil):
(WTF::URLParser::URLParser):
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL const):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::reserveCapacity):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::characters const):
* Source/WTF/wtf/text/StringBuilderJSON.cpp:
(WTF::appendQuotedJSONStringInternal):
(WTF::StringBuilder::appendQuotedJSONString):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::characters<LChar> const): Deleted.
(WTF::String::characters<UChar> const): Deleted.
* Source/WebCore/css/CSSVariableData.cpp:
(WebCore::CSSVariableData::updateBackingStringsInTokens):
* Source/WebCore/dom/make_names.pl:
(printTagNameCppFile):
(printNodeNameCppFile):
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::appendCharactersReplacingEntitiesInternal):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::containsKanaLetters):
* Source/WebCore/html/track/VTTScanner.cpp:
(WebCore::VTTScanner::VTTScanner):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::fixUnparsedProperties):
(WebCore::StyleSheetHandler::fixUnparsedPropertyRanges):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::replaceNonPreservedNewLineAndTabCharactersAndAppend):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::firstUserPerceivedCharacterLength):
* Source/WebCore/rendering/LegacyInlineIterator.h:
(WebCore::LegacyInlineIterator::incrementByCodePointInTextNode):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::xmlDocPtrForString):
* Source/WebGPU/WGSL/Lexer.h:
(WGSL::Lexer::Lexer):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.cpp:
(WebKit::NetworkCache::hexDigitsToHash):
(WebKit::NetworkCache::Key::stringToHash):
* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder<String>::encode):
* Source/WebKit/Shared/API/c/cf/WKStringCF.mm:
(WKStringCopyCFString):

Canonical link: https://commits.webkit.org/278082@main
